packages/contracts-bedrock/src/cannon/MIPS2.sol

```diff
@@ -0,0 +1,642 @@
+            from, to := copyMem(from, to, 32) // memRoot
+            from, to := copyMem(from, to, 32) // preimageKey
+            from, to := copyMem(from, to, 4) // preimageOffset
+            from, to := copyMem(from, to, 4) // heap
+            let exitCode := mload(from)
+            from, to := copyMem(from, to, 1) // exitCode
+            let exited := mload(from)
+            from, to := copyMem(from, to, 1) // exited
+            from, to := copyMem(from, to, 8) // step
+            from, to := copyMem(from, to, 8) // stepsSinceLastContextSwitch
+            from, to := copyMem(from, to, 4) // wakeup
+            from, to := copyMem(from, to, 1) // traverseRight
+            from, to := copyMem(from, to, 32) // leftThreadStack
+            from, to := copyMem(from, to, 32) // rightThreadStack
+            from, to := copyMem(from, to, 4) // nextThreadID
+
+            // Clean up end of memory
+            mstore(to, 0)
+
+            // Log the resulting MIPS state, for debugging
+            log0(start, sub(to, start))
+
+            // Determine the VM status
+            let status := 0
+            switch exited
+            case 1 {
+                switch exitCode
+                // VMStatusValid
+                case 0 { status := 0 }
+                // VMStatusInvalid
+                case 1 { status := 1 }
+                // VMStatusPanic
+                default { status := 2 }
+            }
+            // VMStatusUnfinished
+            default { status := 3 }
+
+            // Compute the hash of the resulting MIPS state and set the status byte
+            out_ := keccak256(start, sub(to, start))
+            out_ := or(and(not(shl(248, 0xFF)), out_), shl(248, status))
+        }
+    }
+
+    /// @notice Updates the current thread stack root via inner thread root in calldata
+    function updateCurrentThreadRoot() internal pure {
+        State memory state;
+        ThreadState memory thread;
+        assembly {
+            state := STATE_MEM_OFFSET
+            thread := TC_MEM_OFFSET
+        }
+        bytes32 updatedRoot = computeThreadRoot(loadCalldataInnerThreadRoot(), thread);
+        if (state.traverseRight) {
+            state.rightThreadStack = updatedRoot;
+        } else {
+            state.leftThreadStack = updatedRoot;
+        }
+    }
+
+    /// @notice Completes the FUTEX_WAIT syscall.
+    function onWaitComplete(
+        State memory _state,
+        ThreadState memory _thread,
+        bool _timeout
+    )
+        internal
+        returns (bytes32 out_)
+    {
+        // Clear the futex state
+        _thread.futexAddr = 0xFF_FF_FF_FF;
+        _thread.futexVal = 0;
+        _thread.futexTimeoutStep = 0;
+
+        // Complete the FUTEX_WAIT syscall
+        _thread.registers[2] = 0;
+        // set errno
+        _thread.registers[7] = _timeout ? sys.ETIMEDOUT : 0;
+        _thread.pc = _thread.nextPC;
+        _thread.nextPC = _thread.nextPC + 4;
```

Comment:
Maybe better to use the syscall helper here:
```suggestion
       v0 = _timeout ? sys.SYS_ERROR_SIGNAL : 0;
       v1  = _timeout ? sys.ETIMEDOUT : 0;
       st.CpuScalars memory cpu = getCpuScalars(thread);
       sys.handleSyscallUpdates(cpu, thread.registers, v0, v1);
       setStateCpuScalars(thread, cpu);
```

Reply:
fixed.